python/kotekan/shared_memory_buffer.py

```diff
@@ -0,0 +1,539 @@
+
+        Parameters
+        ----------
+        n : int
+            Number of time samples to read.
+
+        Returns
+        -------
+        VisRaw
+            Last n time samples.
+
+        Raises
+        ------
+        SharedMemoryError
+            If the shared memory is marked as invalid by the writer or if the structural parameters
+            have changed.
+        """
+        if n > self.buffer_size:
+            raise SharedMemoryReaderUsageError(
+                "Can't read more than {0} last time slots, because buffer size is set to {0} "
+                "(Tried to read last {1} time slots.).".format(self.buffer_size, n)
+            )
+
+        self.shared_mem.seek(0)
+
+        self._validate_shm()
+
+        # get a data update from the ringbuffer
+        access_record = self._access_record()
+
+        times = self._filter_last(access_record, n)
+        logger.debug("Reading last {} time slots: {}".format(n, times))
+
+        self._copy_from_shm(times, access_record)
+
+        # check if any data became invalid while reading it
+        access_record_after_copy = self._access_record()
+        for t in times:
+            for f in range(self.num_freq):
+                if access_record_after_copy[t][f] != access_record[t][f]:
+                    logger.debug(
+                        "Data at t={}, f={} became invalid while reading it.".format(
+                            t, f
+                        )
+                    )
+                    self._data[self._time_index_map[t], f] = None
+        if self._last_access_record is None:
+            self._last_access_record = np.ndarray((self.num_time, self.num_freq))
+        self._last_access_record[times, :] = access_record_after_copy[times, :]
+
+        # return visRaw()
+        return self._return_data_copy_last(n)
+
+    def _remove_oldest_time_slot(self):
+        """
+        Remove the oldest time slot from the time index map.
+
+        Returns
+        -------
+        int or None
+            Index of deleted time slot.
+        """
+        oldest = sorted(self._time_index_map.keys())[-1]
+        return self._time_index_map.pop(oldest)
+
+    def _find_free_time_slot(self):
+        """
+        Find a free time slot in the time index map.
+
+        Returns
+        -------
+        int or None
+            Index of free time slot.
+        """
+        for i in range(self.buffer_size):
+            if i not in self._time_index_map.values():
+                logger.debug("Found free time slot: {}".format(i))
+                return i
+        return None
+
+    def _return_data_copy_last(self, n):
+        if not self._time_index_map:
+            return None
+
+        # sort time index map
+        last_ts, idxs = list(zip(*sorted(self._time_index_map.items())))
+
+        # select last n time slots
+        idxs = idxs[-n:]
+        return VisRaw.from_nparray(
+            self._data[idxs, :],
```

Comment:
This is going to create your third copy of the data, and *must* be able to be avoided. A complex slice like this always generates a copy, and we can't afford the memory. I think the same very much applies for the `read_since` branch.

Reply:
Do you have a suggestion?

Reply:
I'm not completely sure what the best way to do this is. I mentioned this elsewhere, but there's essentially three different reads we'll be doing, of which there are only really two different types:
- Infrequent reads of the whole buffer
- Frequency reads of a small number of recents samples

Currently what you're doing is fine for the second class, but not for the first. If we have a hard requirement that we can't do an extra copy for the first case, then that essentially means you must expose the internal copy.

I think what might work is refactoring along this line.

- First ignore the `read_since` case, it's less important, and not obvious how to fit into what I'm about to say. If we can figure out a way then great, if not, then we drop it.
- Rather than separating what's returned by the `read_*` calls and the internal copy you always return the internal copy directly.
- There is no `.read_last(n)` method, `n` is a parameter supplied at construction, and there is an `.update()` method.
- The internal copy is a ring buffer of length `n` in time. When `.update()` is called, we just update the ring buffer with any new data.
- A (read-only) view of the ring buffer is returned directly by the `.update()` call. This means what the caller gets is a contiguous view of the latest n samples (albeit a cyclic one). You could in theory provide a wrapper which re-indexes the cyclic axis.
- In the basic scenario above you would need to open the buffer again for every type of view of the buffer that you want. This may be done more elegantly by having one class which opens and manages the buffer, and that class has a `.view(n)` method which creates a separate `BufferView` object which holds the internal copy and had the `.update()` method to update it from the shared memory buffer.